java-snapshot-testing-core/src/main/java/au/com/origin/snapshots/reporters/PlainTextSnapshotDiffReporter.java

```diff
@@ -0,0 +1,36 @@
+package au.com.origin.snapshots.reporters;
+
+import au.com.origin.snapshots.exceptions.SnapshotMatchException;
+import org.assertj.core.util.diff.DiffUtils;
+import org.assertj.core.util.diff.Patch;
+
+import java.util.Arrays;
+import java.util.function.Supplier;
+
+public class PlainTextSnapshotDiffReporter implements SnapshotDiffReporter {
+
+    private static final Supplier<IllegalStateException> NO_DIFF_EXCEPTION_SUPPLIER =
+            () -> new IllegalStateException("No differences found. Potential mismatch between comparator and reporter");
+
+    @Override
+    public boolean supportsFormat(String outputFormat) {
+        return true; // always true
+    }
+
+    @Override
+    public void reportDiff(String snapshotName, String rawSnapshot, String currentObject) {
+        Patch<String> patch = DiffUtils.diff(
+                Arrays.asList(rawSnapshot.trim().split("\n")),
+                Arrays.asList(currentObject.trim().split("\n")));
+        throw new SnapshotMatchException("Error on: \n" + currentObject.trim() + "\n\n" + getDiffString(patch));
```

Comment:
If we have multiple reports - I don't think the extra ones will all run if this exception is thrown.  We might need to rethink how and where the exception is throws to allow all reporters to do their work. Perhaps they register themselves as SUCCESS/FAILED and we throw after all have had a change to process the request.

Reply:
Yeah this felt a bit iffy to me too. 

I considered the aggregated reporting method as you suggested but decided to go with what I've done in favour of failing fast and also showing only a limited and concise report and to be able to allow reporters to choose to use their assertion library of choice (assertj, jsonassert for example) so that IDEs can also use that to provide some meaningful diff view (IMO typically much more advanced and interactive than any reporters that we can create)

For what it's worth I added this comment block in the getReporters method in SnapshotConfig to clarify this behaviour. What's your thought on this ? Do you think we should still do the aggregated report at the end ?

    /**
     * Optional
     * Override to supply your own custom reporter functions
     * Reporters will run in the same sequence as provided.
     * Reporters can choose to throw exceptions in which case subsequent reporters will be skipped.
     * Typically one such reporter can be included at the end which can use common assertion libraries
     * like assertj or junit assertions to 'fail' the test. This is useful since IDEs like intellij can
     * then show a diff using their native diff tools.
     * <p>
     * In the absence of any 'throwing' reporters in the list, a default failure exception will be thrown
     * to make sure tests fail.
     *
     * @return custom reporter functions
     */

Reply:
I think ideally we should throw once all reporters get a chance to process the request. Maybe each reporter run gets wrapped in a `try/catch`, we add the exception to a list -  then we re-throw the lot at the end if the errorList is not empty. This may have unintended consequences on the IDE integration you spoke about though.

If you have already taken a look at it and it is problematic - we can look at it later in a separate ticket.

Reply:
I was experimenting earlier today by enhancing SnapshotMatchException to accept a list of Throwables and it wasn't working properly. I was appending all messages separated by System.lineSeparator() but intellij couldn't detect that these were separate errors.

I was trying to see how opentest4j handles this and found something interesting. It seems to have a MultipleFailuresError which takes a list of Throwables and when thrown, intellij is able to identify individual errors and shows a separate diff view for each error 😃 The reason it seems to be working is because it extends AssertionError but SnapshotMatchException extends RuntimeException. 

I think it's reasonable to have SnapshotMatchException extend AssertionError and do something similar to what MultipleFailuresError does so that we can get aggregated reporting. I'll try this out and let you know how it goes.
 
If this works out, I'll also rename SnapshotMatchException to SnapshotMatchError.

Reply:
It only seems to work when I explicitly extend MultipleFailuresError and not AssertionError. 

Related discussion : https://github.com/lukas-krecan/JsonUnit/issues/146 (this library BTW looks promising as a potential reporter)

Extending MultipleFailures from opentest4j doesn't seem ideal (would have preferred if AssertionError worked somehow) but I still think it adds value. Shall I go ahead and make SnapshotMatchException extend MultipleFailuresError ?

Reply:
Yeah - give it a go.

Reply:
What if we copy the code from MultipleFailuresError into this project.  Just don't want an entire library for a single class thats all.

Reply:
Tried that. Didn't work. Looks like IDE handling is specifically done for opentest4j.

I just checked the github page for opentest4j and their goal seems to be to standardize assertions on the JVM and considering that other assertion libraries like assertj and junit assertions also use opentest4j, I feel it's ok to add this dependency.

Reply:
If that's the case we can add it. Perhaps we should we shadow it so it doesn't conflict with other versions like we do for jackson?

Reply:
IDE might not work if we shadow it actually - if it's looking for specific packages etc.

Reply:
Ah that makes sense. Will do.

Just to be clear, I'm guessing you meant relocate and not shadow. Shadowing just makes things even worse 😛 . Will relocate the package.

Reply:
Oh yeah I didn't think about that. Will test it out anyway.

Reply:
Yeah I meant relocate like here: https://github.com/origin-energy/java-snapshot-testing/blob/master/build.gradle#L22-L27